### PR TITLE
Jfrog XRay API parser (summary/artifact) 

### PR DESCRIPTION
**Pull request by madeoninfo**

### CODE REVIEW

File: dojo/tools/jfrog_xray_api_summary_artifact/parser.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    if 'severity' in vulnerability:
+        if vulnerability['severity'] == 'Unknown':
+            severity = "Informational"
+        else:
+            severity = vulnerability['severity'].title()
+    else:
+        severity = "Informational"
+
+    # Some entries have no CVE entries, despite they exist. Example CVE-2017-1000502.
+    cves = vulnerability.get('cves', [])
```
Comment (valentijnscholten) [inline dojo/tools/jfrog_xray_api_summary_artifact/parser.py:64]: you can now set multiple identifiers in the vulnerability references relation

File: dojo/tools/jfrog_xray_api_summary_artifact/parser.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    else:
+        title = "No CVE - " + impact_path.name + ":" + impact_path.version
+
+    unique_id_from_tool = title
+
+    finding = Finding(
+        service=service,
+        title=title,
+        cwe=cwe,
+        cve=cve,
```
Comment (valentijnscholten) [inline dojo/tools/jfrog_xray_api_summary_artifact/parser.py:92]: I think 'No CVE' can end up here which is an invalid value

File: dojo/tools/jfrog_xray_api_summary_artifact/parser.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        cvssv3=cvssv3,
+        cvssv3_score=cvssv3_score,
+        severity=severity,
+        description=vulnerability['description'],
+        verified=False,
```
Comment (valentijnscholten) [inline dojo/tools/jfrog_xray_api_summary_artifact/parser.py:97]: this field should not be set by a parser

File: dojo/tools/jfrog_xray_api_summary_artifact/parser.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        test=test,
+        hash_code=impact_path.sha,
```
Comment (valentijnscholten) [inline dojo/tools/jfrog_xray_api_summary_artifact/parser.py:99]: this field should not be set by a parser

File: dojo/tools/jfrog_xray_api_summary_artifact/parser.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        impact_path = decode_impact_path(impact_paths[0])
+
+    if 'issue_id' in vulnerability:
+        title = vulnerability['issue_id'] + " - " + impact_path.name + ":" + impact_path.version
+    elif cve:
+        title = str(cve) + " - " + impact_path.name + ":" + impact_path.version
+    else:
+        title = "No CVE - " + impact_path.name + ":" + impact_path.version
+
+    unique_id_from_tool = title
```
Comment (valentijnscholten) [inline dojo/tools/jfrog_xray_api_summary_artifact/parser.py:86]: do not set this field if there is no unique value in the actual report

**Review by valentijnscholten — CHANGES_REQUESTED** [review]
thanks for the PR, left some comments

File: dojo/tools/jfrog_xray_api_summary_artifact/parser.py (hunk 1 above)
Comment (madeoninfo) [reply dojo/tools/jfrog_xray_api_summary_artifact/parser.py:64]: Hi, not sure what you mean by this comment.

File: dojo/tools/jfrog_xray_api_summary_artifact/parser.py (hunk 5 above)
Comment (madeoninfo) [reply dojo/tools/jfrog_xray_api_summary_artifact/parser.py:88]: Replaced with the sha for the uniqueness of the title. Is there a limit though on the length of the title?

File: dojo/tools/jfrog_xray_api_summary_artifact/parser.py (hunk 4 above)
Comment (madeoninfo) [reply dojo/tools/jfrog_xray_api_summary_artifact/parser.py]: Removed

File: dojo/tools/jfrog_xray_api_summary_artifact/parser.py (hunk 3 above)
Comment (madeoninfo) [reply dojo/tools/jfrog_xray_api_summary_artifact/parser.py]: Removed

File: dojo/tools/jfrog_xray_api_summary_artifact/parser.py (hunk 2 above)
Comment (madeoninfo) [reply dojo/tools/jfrog_xray_api_summary_artifact/parser.py:94]: Initialization of cve chenged from "No CVE" to "none"

File: dojo/tools/jfrog_xray_api_summary_artifact/parser.py (hunk 5 above)
Comment (valentijnscholten) [reply dojo/tools/jfrog_xray_api_summary_artifact/parser.py:88]: check models.py

File: dojo/tools/jfrog_xray_api_summary_artifact/parser.py (hunk 1 above)
Comment (valentijnscholten) [reply dojo/tools/jfrog_xray_api_summary_artifact/parser.py:64]: You can add all CVEs from the report to the `vulnerability_reference_set` field of a Finding, check https://github.com/DefectDojo/django-DefectDojo/pull/5904

Comment (madeoninfo) [reply dojo/tools/jfrog_xray_api_summary_artifact/parser.py:64]: Having deployment issues at this point so I can't test this on my machine. Added a question on the [defectdojo slack channel](https://owasp.slack.com/archives/C2P5BA8MN/p1650980167639499). Could we complete this PR and I will create another one once I'm up and running again?

File: dojo/tools/jfrog_xray_api_summary_artifact/parser.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        if vulnerability['severity'] == 'Unknown':
+            severity = "Informational"
+        else:
+            severity = vulnerability['severity'].title()
+    else:
+        severity = "Informational"
+
+    # Some entries have no CVE entries, despite they exist. Example CVE-2017-1000502.
+    cves = vulnerability.get('cves', [])
+    if len(cves) > 0:
```
Comment (damiencarol) [inline dojo/tools/jfrog_xray_api_summary_artifact/parser.py:65]:
```suggestion
    if cves:
```

File: dojo/tools/jfrog_xray_api_summary_artifact/parser.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+        return items
+
+
+# Retrieve the findings
+def get_item(vulnerability, service, test):
+    cve = None
+    cwe = None
+    cvssv3 = None
+    cvss_v3 = "No CVSS v3 score"
```
Comment (damiencarol) [inline dojo/tools/jfrog_xray_api_summary_artifact/parser.py:49]:
```suggestion
```

File: dojo/tools/jfrog_xray_api_summary_artifact/parser.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    cves = vulnerability.get('cves', [])
+    if len(cves) > 0:
+        if 'cve' in cves[0]:
+            cve = cves[0]['cve']
+        if len(cves[0].get('cwe', [])) > 0:
+            cwe = decode_cwe_number(cves[0].get('cwe', [])[0])
+        if 'cvss_v3' in cves[0]:
+            cvss_v3 = cves[0]['cvss_v3']
+            cvssv3 = CVSS3.from_rh_vector(cvss_v3).clean_vector()
+            cvssv3_score = CVSS3.from_rh_vector(cvss_v3).base_score
```
Comment (damiencarol) [inline dojo/tools/jfrog_xray_api_summary_artifact/parser.py:73]:
```suggestion
```
This is automatic, parser should focus on the vector if present.

**Review by damiencarol — CHANGES_REQUESTED** [review]

**Comment by damiencarol** [comment]
@madeoninfo pushing a first build but the comment of @valentijnscholten is still valid. if you can please leverage the new vulne reference attribute.

**Comment by madeoninfo** [comment]
Hi @damiencarol and @valentijnscholten. I'm still having issues with DefectDojo on Windows with Docker Desktop. Got a 500 error with an empty database, even though the initializer has completed its task. I am setting up a new dev environment on Linux as Docker Desktop might be the culprit with this. I promise to supply an update on the comments and also include more parsers related to JFrog from the API that they provide.

**Comment by madeoninfo** [comment]
Hi @damiencarol, without a setup localy it is difficult to test and I am relying on the unittests of the PR to verify the functionality. Could you please run them to make sure I have not broken anything?

**Review by damiencarol — CHANGES_REQUESTED** [review]
You have 2 times the parser defined. (see __init__.py)

**Comment by madeoninfo** [comment]
> You have 2 times the parser defined. (see **init**.py)

Could you please point this out in the code?

File: dojo/tools/jfrog_xray_api_summary_artifact/parser.py [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    else:
+        title = impact_path.name + ":" + impact_path.version
+
+    finding = Finding(
+        service=service,
+        title=title,
+        cwe=cwe,
+        cve=cve,
+        cvssv3=cvssv3,
+        cvssv3_score=cvssv3_score,
```
Comment (damiencarol) [inline dojo/tools/jfrog_xray_api_summary_artifact/parser.py:90]:
```suggestion
```
The score is automatically calculated. Better to not set it if the tool doesn't have the data

Comment (madeoninfo) [reply dojo/tools/jfrog_xray_api_summary_artifact/parser.py]: Removed

**Comment by madeoninfo** [comment]
Making a few changes to the unique_id_from_tool and will push soon

**Review by damiencarol — APPROVED** [review]

**Review by Maffooch — APPROVED** [review]

**Pull request merged**